### PR TITLE
Fix validation and sanitization of application method

### LINKED ISSUE #1814: Submitting existing job doesn't pre-fill with posted values
WP Job Manager will use posted values to pre-fill form data in case when listing is new, but it doesn't seem to do that when editing existing listing, and instead pre-fills the form data with existing listing data (if a server side validation error occurs). 

#### Steps to Reproduce
1. Visit job dashboard, edit existing listing
2. Make a change to any field
3. Bypass client side validation (html5 required attribute) for some required field and make that field empty to trigger server side validation error
4. Any changes made in 2. are lost

#### WordPress / WP Job Manager Version
Reproducible on https://demo.wpjobmanager.com

#### Screenshot / Video
https://www.dropbox.com/s/ccvdsilgaqvjzvp/2019-07-01_04-57-22.mp4?dl=0

### LINKED ISSUE #2067: "Application email/URL" expects the the URL protocol but is not clear in the error or description
The "Application email/URL" job submission field expects the protocol (HTTP:// or HTTPS://) when entering a URL. However this is not explicitly stated nor is it clear in the error message when one does not enter the protocol. 

The only error shown is `Application email/URL is a required field`. 

We should make it clearer by using description, placeholder text, and/or error message that the protocol is required.

3415016-zen

### PR DESCRIPTION
**Pull request by jom**
Fixes #2067
Fixes #1814
Replaces #1829

### Changes proposed in this Pull Request

* Adds schema (http or https, depending on filter) to URL application methods when possible/necessary.
* Allows sanitization errors to show on required fields. To do this, added an extra `before_sanitize` callback property on the field. Implemented internally but optional to implement (default is based on emptiness of sanitized value) for custom fields.
* Invalid emails are cleared.
* Note: saving to drafts will allow invalid data to be saved, but not non-sanitized data. When job is properly submitted, validation will occur. 
* Eventually we should add field level error handling. Too large of task for this PR, though.

### Testing instructions

* Using all setting variations for Application Method (email, url, both), fill out the job submission form. 
* Fill out the URL/both frontend application form field with various invalid URLS (note, `filter_var` is pretty relaxed in its URL validation rules). Try just putting "google.com" and verifying it prefixes it with the schema.
* Fill out the email/both frontend application form field with various invalid emails. This is a bit more strict... for example `()@example.com` shouldn't work. 

### New/Updated Hooks

* `job_manager_form_assumed_url_scheme`: Allow setting some URLs to a `https` schema by default when no schema set.

### CODE REVIEW

File: includes/abstracts/abstract-wp-job-manager-form.php [hunk 1]
```diff
@@ -651,10 +732,16 @@ protected function upload_file( $field_key, $field ) {
...
 			if ( ! empty( $field['multiple'] ) ) {
-				return $file_urls;
+				$value = $file_urls;
 			} else {
-				return current( $file_urls );
+				$value = current( $file_urls );
 			}
 		}
+
+		if ( isset( $field['before_sanitize'] ) && is_callable( $field['before_sanitize'] ) ) {
```
Comment (jom) [inline includes/abstracts/abstract-wp-job-manager-form.php:741]: TBH this doesn't do much and is called in a weird spot, but just thought I'd add it for consistency.

File: includes/forms/class-wp-job-manager-form-submit-job.php [hunk 2]
```diff
@@ -358,7 +358,11 @@ protected function should_application_field_skip_email_url_validation() {
 	protected function validate_fields( $values ) {
 		foreach ( $this->fields as $group_key => $group_fields ) {
 			foreach ( $group_fields as $key => $field ) {
-				if ( $field['required'] && empty( $values[ $group_key ][ $key ] ) ) {
+				if (
+					$field['required']
+					&& empty( $values[ $group_key ][ $key ] )
+					&& ( ! isset( $field['empty'] ) || $field['empty'] )
```
Comment (jom) [inline includes/forms/class-wp-job-manager-form-submit-job.php:364]: This makes the `empty` parameter optional if set by the posted field callback.

File: includes/forms/class-wp-job-manager-form-submit-job.php [hunk 3]
```diff
@@ -451,31 +455,32 @@ protected function validate_fields( $values ) {
...
 		// Application method.
-		if ( ! $this->should_application_field_skip_email_url_validation() && isset( $values['job']['application'] ) && ! empty( $values['job']['application'] ) ) {
+		if ( ! $this->should_application_field_skip_email_url_validation() && isset( $values['job']['application'] ) ) {
 			$allowed_application_method   = get_option( 'job_manager_allowed_application_method', '' );
-			$values['job']['application'] = str_replace( ' ', '+', $values['job']['application'] );
+
+			$is_valid = true;
+
+			// phpcs:ignore WordPress.Security.NonceVerification.Missing -- Nonce checked earlier when required.
+			$posted_value = isset( $_POST['application'] ) ? sanitize_text_field( wp_unslash( $_POST['application'] ) ) : false;
```
Comment (jom) [inline includes/forms/class-wp-job-manager-form-submit-job.php:464]: Relaxing the sanitization for the re-submit just so it isn't quite as jarring for the user.

File: includes/abstracts/abstract-wp-job-manager-form.php [hunk 4]
```diff
@@ -432,6 +432,18 @@ protected function get_posted_fields() {
...
 
+				$this->fields[ $group_key ][ $key ]['empty'] = false;
+
+				if ( ! isset( $field['before_sanitize'] ) ) {
+					$field['before_sanitize'] = function ( $value ) use ( $group_key, $key ) {
+						if ( is_string( $value ) ) {
+							$value = trim( $value );
+						}
+
+						$this->fields[ $group_key ][ $key ]['empty'] = empty( $value );
```
Comment (gkaragia) [inline includes/abstracts/abstract-wp-job-manager-form.php:443]: Won't this report more values that it should as empty? We were checking for null only previously.

**Review by gkaragia — COMMENTED** [review]
Looks good! There might be a possible issue.

File: includes/abstracts/abstract-wp-job-manager-form.php (hunk 4 above)
Comment (jom) [reply includes/abstracts/abstract-wp-job-manager-form.php:443]: This is essentially just an earlier check from what happens in `\WP_Job_Manager_Form_Submit_Job::validate_fields` for the `x is a required field`, which uses and has always used just `empty()`.

Comment (gkaragia) [reply includes/abstracts/abstract-wp-job-manager-form.php:443]: Cool! Missed that!

**Review by gkaragia — APPROVED** [review]
Looks good, works well!

**Pull request merged**